Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/ChangeFeedProcessorCore.cs

```diff
@@ -75,23 +80,99 @@ public override async Task StopAsync()
             Routing.PartitionKeyRangeCache partitionKeyRangeCache = await this.monitoredContainer.ClientContext.DocumentClient.GetPartitionKeyRangeCacheAsync(NoOpTrace.Singleton);
             if (this.documentServiceLeaseStoreManager == null)
             {
-                this.documentServiceLeaseStoreManager = await DocumentServiceLeaseStoreManagerBuilder.InitializeAsync(this.monitoredContainer, this.leaseContainer, leaseContainerPrefix, this.instanceName).ConfigureAwait(false);
+                this.documentServiceLeaseStoreManager = await DocumentServiceLeaseStoreManagerBuilder
+                    .InitializeAsync(
+                        this.monitoredContainer,
+                        this.leaseContainer,
+                        leaseContainerPrefix,
+                        this.instanceName,
+                        changeFeedMode: this.changeFeedProcessorOptions.Mode)
+                    .ConfigureAwait(false);
             }
 
+            await this.ChangeFeedModeSwitchingCheckAsync().ConfigureAwait(false);
+
             this.partitionManager = this.BuildPartitionManager(
-                containerRid, 
+                containerRid,
                 partitionKeyRangeCache);
             this.initialized = true;
         }
 
+        /// <summary>
+        /// If the lease container's lease document is found, this method checks for lease 
+        /// document's ChangeFeedMode and if the new ChangeFeedMode is different
+        /// from the current ChangeFeedMode, a CosmosException is thrown.
+        /// This is based on an issue located at <see href="https://github.com/Azure/azure-cosmos-dotnet-v3/issues/4308"/>.
+        /// </summary>
+        private async Task ChangeFeedModeSwitchingCheckAsync()
+        {
+            IReadOnlyList<DocumentServiceLease> documentServiceLeases = await this.documentServiceLeaseStoreManager
+                .LeaseContainer
+                .GetAllLeasesAsync()
+                .ConfigureAwait(false);
+
+            // No lease documents. Return.
+
+            if (documentServiceLeases.Count == 0)
+            {
+                return;
+            }
+
+            DocumentServiceLease documentServiceLease = documentServiceLeases.FirstOrDefault();
+
+            // No lease documents that match the Id.
+
+            if (documentServiceLease == default)
+            {
+                return;
+            }
+
+            // Mode attribute exists on lease document, but it is not set. legacy is always LatestVersion because
+            // AllVersionsAndDeletes does not exist. There should not be any legacy lease documents that are
+            // AllVersionsAndDeletes. If the ChangeFeedProcessor's mode is not legacy, a CosmosException should thrown.
+            // If the ChangeFeedProcessor mode is not the mode in the lease document, a CosmosException should be thrown.
+
+            bool shouldThrowException = string.IsNullOrEmpty(documentServiceLease.Mode)
```

Comment:
nit: This could be a single line:

```suggestion
            bool shouldThrowException = this.VerifyChangeFeedProcessorMode(
                    changeFeedMode: string.IsNullOrEmpty(documentServiceLease.Mode) ? ChangeFeedMode.LatestVersion : documentServiceLease.Mode,
                    leaseChangeFeedMode: documentServiceLease.Mode,
                    normalizedProcessorChangeFeedMode: out string normalizedProcessorChangeFeedMode);
```

Reply:
since it is a nit, i will resolve, unless you feel strongly about it, then remove nit and we can talk offline?

Reply:
No strong feelings, I was just mentioning it to avoid the repeated lines for the other parameters

Reply:
Repeating the lines leads to things like `out string normalizedProcessorChangeFeedMode` and `out normalizedProcessorChangeFeedMode` on the other because you cannot add `string` to both

Reply:
What I just committed will hopefully be a good compromise. @ealsur